addons/pos_loyalty/models/loyalty_program.py

```diff
@@ -52,11 +52,45 @@ def _compute_pos_config_ids(self):
                 program.pos_config_ids = False
 
     def _compute_pos_order_count(self):
-        read_group_res = self.env['pos.order.line']._read_group(
-            [('reward_id', 'in', self.reward_ids.ids)], ['reward_id:array_agg'], ['order_id'])
-        for program in self:
-            program_reward_ids = program.reward_ids.ids
-            program.pos_order_count = sum(1 if any(id in group['reward_id'] for id in program_reward_ids) else 0 for group in read_group_res)
+        query = """
+                WITH
+                reward_pos_order_matching AS (
+                    SELECT
+                        count(DISTINCT po.id) as pos_order_ids,
+                        lr.id as loyalty_reward_id
+                    FROM
+                        pos_order_line pos_order_line
+                        JOIN pos_order po ON pos_order_line.order_id = po.id
+                        JOIN loyalty_reward lr ON pos_order_line.reward_id IS NOT NULL
+                        AND pos_order_line.reward_id = lr.id
+                    GROUP BY
+                        loyalty_reward_id
+                ),
+                program_reward_matching AS (
+                    SELECT
+                        lr.id as loyalty_reward_id,
+                        lp.id as loyalty_program_id
+                    FROM
+                        loyalty_program lp
+                        JOIN loyalty_reward lr ON lr.program_id = lp.id
+                    WHERE
+                        lp.id = ANY(%s)
+                )
+            SELECT
+                program_reward_matching.loyalty_program_id,
+                SUM(reward_pos_order_matching.pos_order_ids)
+            FROM
+                program_reward_matching program_reward_matching
+                LEFT JOIN reward_pos_order_matching reward_pos_order_matching ON program_reward_matching.loyalty_reward_id = reward_pos_order_matching.loyalty_reward_id
+            GROUP BY
+                program_reward_matching.loyalty_program_id
```

Comment:
```suggestion
         WITH reward_to_orders_count AS (
                 SELECT reward.id                    AS reward_id, 
                        COUNT(DISTINCT pos_order.id) AS orders_count
                   FROM pos_order_line line
                   JOIN pos_order ON line.order_id = pos_order.id
                   JOIN loyalty_reward reward ON line.reward_id = reward.id
               GROUP BY reward_id
              ),
              program_to_reward AS (
                 SELECT reward.id  AS reward_id,
                        program.id AS program_id
                   FROM loyalty_program program
                   JOIN loyalty_reward reward ON reward.program_id = program.id
                  WHERE program.id = ANY (%s)
              )
       SELECT program_to_reward.program_id,
              SUM(COALESCE(reward_to_orders_count.orders_count, 0))
         FROM program_to_reward
    LEFT JOIN reward_to_orders_count ON reward_to_orders_count.reward_id = program_to_reward.reward_id
     GROUP BY program_to_reward.program_id;
```

Sorry I had forgotten to tell you to use the Tabular Right alignment, it's a good 99% to what's called the "River" style SQL guide, to have readable SQL code (notice the blank space between the keyword and code form a sort of 'river', hence the name) This is style you will see enforced in the Upgrade repo and some R&D teams also enforces it. It's a good practice to get used to it. (starting too myself also).

Regardless of formatting, here are some points I've addressed:
- aliases are not needed if they have the same name as the table, the point of an alias is to give it a *meaningful* name that isn't too long that would make the query to verbose.
- renamed the CTEs names so they represent more what they are returning, a first reader can have a better idea what it is doing, without parsing the subquery.
- fixed some of the aliases, to avoid 2 letter aliases, also removed the `loyalty_` prefix in a few places, it doesn't provide any more value as information, just makes the query more verbose.
- I've removed a redundant clause in one of the JOIN in the first CTE, as a `NOT NULL` check here is useless if you are also joining based on a value that not nullable (the Pkey of another table).
- Preferably (nitpick), the left-hand-side of a JOIN condition should be based on the joined table (so the table on the same line usually) and external table should go on the right-hand-side. It helps with reading.

⚠️ Potential problem that needs to be addressed:
In the main query, you are doing a LEFT JOIN, but you are summing on a column that comes from that joined table. Since it's a LEFT JOIN, it's potentially NULL, therefor in the sum you should COALESCE to 0 in case order_counts is NULL. In your case it might not be a problem in general, as NULL entries in a SUM routine are ignored when summing, but if all entries are NULL, the total SUM will be NULL, not 0, this might not be what the external code is expecting, and it's error prone.

Reply:
About the null values, such cases are handled during the assignation loop where if the value is NULL, the value is defaulted to 0 here -->
https://github.com/odoo/odoo/blob/ae19ef58046d7a9fbd0bd72abc4bb77cddab4c21/addons/pos_loyalty/models/loyalty_program.py#L82